File: ros2_ws/src/my_python_pkg/my_python_pkg/add_two_ints_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.srv import AddTwoInts
from functools import partial
 
class AddTwoIntsClient(Node): 
    def __init__(self):
        super().__init__("add_two_ints_client") 
        self.client = self.create_client(AddTwoInts, "add_two_ints")

    def call_add_two_ints(self, a: int, b: int):
        # Until Server is not found Up and Running
        while not self.client.wait_for_service(timeout_sec=1.0):
            self.get_logger().warning("Service not available, waiting again...")

        # When Server is Up and Running, Create Request
        request = AddTwoInts.Request()
        request.a = a
        request.b = b

        # Make Asynchronous Call to Server via Client and wait for the response to come.
        future = self.client.call_async(request)
        # Once future is complete, call the callback function to process the response

        ## IMPORTANT ## Use partial to pass the request object to the callback function, so that we can access the request 
        # parameters in the callback. This is because the callback function only receives the future object as an argument,
        # and we need to access the request parameters to log them in the callback.

        future.add_done_callback(partial(self.callback_call_add_two_ints, request=request))

    def callback_call_add_two_ints(self, future, request):
        response = future.result()
        self.get_logger().info(f"Request params were {request.a}, {request.b}, Client GOT Result of {response.sum}")


def main(args=None):
    rclpy.init(args=args)
    node = AddTwoIntsClient() 
    node.call_add_two_ints(10, 20)
    node.call_add_two_ints(30, 40)
    node.call_add_two_ints(50, 60)
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: ros2_ws/src/my_cpp_pkg/src/add_two_ints_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/srv/add_two_ints.hpp" 


class AddTwoIntsClientNode : public rclcpp::Node
{
public:
    AddTwoIntsClientNode() : Node("add_two_ints_client")  {
    client_ = this->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");

    }

    void call_add_two_ints(int a, int b) {
        // Wait for the service to be available if no service found after 1sec.
        while (!client_->wait_for_service(std::chrono::seconds(1))) {
            RCLCPP_WARN(this->get_logger(), "Service not available, waiting again...");
        }

        //When Server is Up and Running, create a request object and fill the request data.
        auto request = std::make_shared<example_interfaces::srv::AddTwoInts::Request>();
        request->a = a;
        request->b = b;

        // Make Asynchronous Call here is NOT done here...we use a callback to process the response when it is received. This is a better way to do it than blocking the node.
        client_->async_send_request(request, std::bind(&AddTwoIntsClientNode::callback_add_two_ints, this, std::placeholders::_1)); //1 parameter here, cant use partial() like python here. 
        // Also, here this is not the value in future, future is the  param name in the callbacl, the value received here is assigned to future once callback is triggered.
    }

    void callback_add_two_ints(rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedFuture future) {
        auto response = future.get();
        RCLCPP_INFO(this->get_logger(), "Result: %ld", response->sum);
    }
 
private:
    rclcpp::Client<example_interfaces::srv::AddTwoInts>::SharedPtr client_; // can also write auto.

};
 
int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<AddTwoIntsClientNode>(); 
    node->call_add_two_ints(56, 44);
    node->call_add_two_ints(100, 200);
    node->call_add_two_ints(10, 20);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}